### PR TITLE
HDMA

### PR DESCRIPTION
**Pull request by Valgard**

### CODE REVIEW

File: yagb-core/src/emulator/ppu/ppu-base.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { Bus, ReadHandler, WriteHandler } from '../bus';
 import { Interrupt, irq } from '../interrupt';
 import { Ppu, ppuMode } from '../ppu';
 
+import { Event } from 'microevent.ts';
```
Comment (DirtyHairy) [inline yagb-core/src/emulator/ppu/ppu-base.ts:5]: I always happy to see `microevent.ts` in use, but I think in this case we'd be better served with a method called `onHblankStart` which is empty in `PpuBase` and overridden in `PpuGbc` --- less code, and slightly less overhead.

File: yagb-core/src/emulator/ppu/ppu-cgb.ts [hunk 2]
```diff
@@ -389,6 +405,101 @@ export class PpuCgb extends PpuBase {
         }
     };
 
+    private invalidRead: ReadHandler = (address) => {
+        this.system.log(`Invalid read from HDMA register ${hex16(address)}`);
```
Comment (DirtyHairy) [inline yagb-core/src/emulator/ppu/ppu-cgb.ts:409]: The logs are fine now, but I'd like to remove them once we have CGB working.

File: yagb-core/src/emulator/ppu/ppu-cgb.ts [hunk 3]
```diff
@@ -389,6 +405,101 @@ export class PpuCgb extends PpuBase {
...
+        this.resetHDMA();
+    };
+
+    private hdmaCopyMemory(length: number) {
+        const destination = 0x8000 + this.hdmaDestination;
+
+        let value;
+        for (let address = 0; address < length; address++) {
+            if (0x9ff0 < destination + address) {
+                this.system.log('Address overflow during HDMA transfer');
```
Comment (DirtyHairy) [inline yagb-core/src/emulator/ppu/ppu-cgb.ts:471]: I am pretty sure that what actually happens is that the address wraps back to the beginning of VRAM. Maybe do that instead, and keep the log while developmet

File: yagb-core/src/emulator/ppu/ppu-cgb.ts [hunk 4]
```diff
@@ -389,6 +405,101 @@ export class PpuCgb extends PpuBase {
...
+    };
+
+    private hdma5Read: ReadHandler = () => {
+        if (this.hdmaLength === 0) return 0xff;
+
+        return (this.hdmaOn ? 0x00 : 0x80) | ((this.hdmaLength >> 4) - 1);
+    };
+
+    private hdma5Write: WriteHandler = (_, value) => {
+        if (this.hdmaOn && TransferMode.hblank === this.hdmaMode) {
```
Comment (DirtyHairy) [inline yagb-core/src/emulator/ppu/ppu-cgb.ts:441]: According to the cycle accurate gameboy doc, HDMA does not execute if the display is switched off. The correct behaviour is slightly is even more difficult, but I think returning if the LCD is off would be sufficient for now.

File: yagb-core/src/emulator/ppu/ppu-cgb.ts [hunk 5]
```diff
@@ -389,6 +405,101 @@ export class PpuCgb extends PpuBase {
...
+
+        this.hdmaCopyMemory(this.hdmaLength);
+
+        this.resetHDMA();
+    };
+
+    private hdmaCopyMemory(length: number) {
+        const destination = 0x8000 + this.hdmaDestination;
+
+        let value;
```
Comment (DirtyHairy) [inline yagb-core/src/emulator/ppu/ppu-cgb.ts:468]: I for one would prefer if the `let` were moved into the loop body.

File: yagb-core/src/emulator/ppu/ppu-cgb.ts (hunk 3 above)
Comment (Valgard) [reply yagb-core/src/emulator/ppu/ppu-cgb.ts:471]: From Pandoc:
If the transfer’s destination address overflows, the transfer stops prematurely. The status of the registers if this happens still needs to be investigated.
https://gbdev.io/pandocs/CGB_Registers.html#bit-7--1--hblank-dma

**Pull request merged**